internal/k8sbridge/clientset.go

```diff
@@ -1,44 +1,128 @@
 package k8sbridge
 
 import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	k8schema "k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/client-go/kubernetes"
-	clientscheme "k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/rest"
+
+	"github.com/grafana/grafana/pkg/schema"
+)
+
+var (
+	// ErrCRDAlreadyRegistered is returned when trying to register a duplicate CRD.
+	ErrCRDAlreadyRegistered = errors.New("error registering duplicate CRD")
 )
 
-// Clientset
+// Clientset is the clientset for Kubernetes APIs.
+// It provides functionality to talk to the APIs as well as register new API clients for CRDs.
 type Clientset struct {
+	// TODO: this needs to be exposed, but only specific types (e.g. no pods / deployments / etc.).
 	*kubernetes.Clientset
-	// This could be its own type, but I don't see that as necessary ATM.
-	grafanacorev1client *rest.RESTClient
+	extset *apiextensionsclient.Clientset
+	config *rest.Config
+	crds   map[k8schema.GroupVersion]apiextensionsv1.CustomResourceDefinition
+	lock   sync.RWMutex
 }
 
-// NewClientset
+// NewClientset returns a new Clientset configured with cfg.
 func NewClientset(cfg *rest.Config) (*Clientset, error) {
-	clientset, err := kubernetes.NewForConfig(cfg)
+	k8sset, err := kubernetes.NewForConfig(cfg)
 	if err != nil {
 		return nil, err
 	}
 
-	c := *cfg
-	c.NegotiatedSerializer = clientscheme.Codecs.WithoutConversion()
-	c.GroupVersion = &k8schema.GroupVersion{
-		Group:   groupName,
-		Version: groupVersion,
-	}
-	cli, err := rest.RESTClientFor(&c)
+	extset, err := apiextensionsclient.NewForConfig(cfg)
 	if err != nil {
 		return nil, err
 	}
 
 	return &Clientset{
-		clientset,
-		cli,
+		Clientset: k8sset,
+		extset:    extset,
+		config:    cfg,
+		crds:      make(map[k8schema.GroupVersion]apiextensionsv1.CustomResourceDefinition),
 	}, nil
 }
 
-// GrafanaCoreV1
-func (c *Clientset) GrafanaCoreV1() *rest.RESTClient {
-	return c.grafanacorev1client
+// RegisterSchema registers a new client and CRD for schema s.
+func (c *Clientset) RegisterSchema(ctx context.Context, s schema.ObjectSchema) error {
+	ver := k8schema.GroupVersion{
+		Group:   s.GroupName(),
+		Version: s.GroupVersion(),
+	}
+
+	c.lock.RLock()
+	_, ok := c.crds[ver]
+	c.lock.RUnlock()
+	if ok {
+		return ErrCRDAlreadyRegistered
+	}
+
+	crdObj := newCRD(s.Name(), s.GroupName(), s.GroupVersion(), s.OpenAPISchema())
+	crd, err := c.extset.
+		ApiextensionsV1().
+		CustomResourceDefinitions().
+		Create(ctx, &crdObj, metav1.CreateOptions{})
+	if err != nil && !kerrors.IsAlreadyExists(err) {
+		return err
+	}
+
+	c.lock.Lock()
+	c.crds[ver] = *crd
+	c.lock.Unlock()
+
+	return nil
+}
+
+func newCRD(
```

Comment:
I'm gonna guess that, at least for Thema-based schemas (and let's hope that's all we end up with), we can do this generically in thema/crd, using this: https://github.com/grafana/thema/blob/main/crd/crd.cue

Getting that framework together will also allow proper control over things like name/pluralization, per the TODOs below.

Big question remains around where `served` and `storage` should come from, though.

Reply:
Yeah, that's the idea here. As far as `served` / `storage` goes, I think we need to consider this in the scope of multiple versions, which (I assume) would be supported by lineages.